### PR TITLE
execute: add an end-to-end --link mode test

### PR DESCRIPTION
**Pull request by marksliva**

### CODE REVIEW

File: test/helpers.bash [hunk 1]
```diff
@@ -51,3 +51,19 @@ delete_cluster() {
     # data directory to segments with "file exists".
     rm -rf "${datadir}"/*_upgrade
 }
+
+# require_gnu_stat tries to find a GNU stat program. If one is found, it will be
+# assigned to the STAT global variable; otherwise the current test is skipped.
+require_gnu_stat() {
```
Comment (jchampio) [inline test/helpers.bash]: I think there's a second copy of this in `checks.bats` that should be removed.

File: test/execute.bats [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    ")
+
+    for relfilenode in "${relfilenodes[@]}"; do
+        local number_of_hardlinks=$($STAT --format "%h" "${relfilenode}")
+        [ $number_of_hardlinks -eq $expected_number_of_hardlinks ] \
+            || fail "expected $expected_number_of_hardlinks hardlinks to $relfilenode; found $number_of_hardlinks"
+    done
+}
+
+set_primary_datadirs() {
```
Comment (jchampio) [inline test/execute.bats]: nit: `primary` makes it seem like we're excluding the master, but we're not.

File: test/execute.bats [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        --verbose
+
+    local datadir=$(dirname $(dirname "${MASTER_DATA_DIRECTORY}"))
+    NEW_CLUSTER="${datadir}/qddir_upgrade/demoDataDir-1"
+
+    gpupgrade execute --verbose
+
+    PGPORT=50432 ensure_hardlinks_for_relfilenode_on_master_and_segments 'test_linking' 2
+
+    reset_primary_pg_control_files
```
Comment (jchampio) [inline test/execute.bats]: nit: any way to move this into the `teardown` somehow? That way we still have a working source cluster if the test fails.

Comment (berlin-ab) [reply test/execute.bats]: The `TEARDOWN_FUNCTIONS`  pattern we introduced might be useful here.

Comment (marksliva) [reply test/execute.bats]: Done.

File: test/execute.bats (hunk 2 above)
Comment (marksliva) [reply test/execute.bats]: Changed it to include `master`.

File: test/helpers.bash (hunk 1 above)
Comment (marksliva) [reply test/helpers.bash]: Done.

**Comment by marksliva** [comment]
> I think my only blocking concern is the second copy of `require_gnu_stat` that needs to be removed.

Done.

**Pull request merged**